### PR TITLE
Add support for external reward functions

### LINKED ISSUE #358: Reward function interface and dynamic loading/binding of implementation

### LINKED ISSUE #359: Reward functions are editable external scripts

### PR DESCRIPTION
**Pull request by phillipleblanc**
Closes #359 
Closes #358

### CODE REVIEW

File: pkg/pods/pod.go [hunk 1]
```diff
@@ -273,6 +275,29 @@ func (pod *Pod) ActionsArgs() []string {
...
+	}
+
+	if !strings.HasSuffix(pod.Training.RewardFuncs, ".py") {
+		return "", errors.New("external reward functions must be defined in a single Python file - see https://docs.spiceai.org/concepts/rewards/")
+	}
+
+	// If we've already loaded the file, just return the contents
+	if pod.externalRewardMemo != "" {
+		return pod.externalRewardMemo, nil
+	}
```
Comment (lukekim) [inline pkg/pods/pod.go]: Could probably just put this check as the first thing in the func

Comment (lukekim) [reply pkg/pods/pod.go]: Actually, I would move all this code to loadPod and just return pod.externalRewardMemo here. Let's move to this pattern for runtime performance.

**Review by lukekim — DISMISSED** [review]

File: ai/src/exec.py [hunk 2]
```diff
@@ -8,3 +10,11 @@ def somewhat_safe_exec(code: str, locals_dict: dict) -> dict:
 
 def somewhat_safe_eval(code: str, locals_dict: dict) -> any:
     return eval(code, {"__builtins__": {}}, locals_dict)  # pylint: disable=eval-used
+
+
+def load_module_from_code(code: str, module_name: str):
+    spec = importlib.util.spec_from_loader(module_name, loader=None)
+    module = importlib.util.module_from_spec(spec)
+
+    exec(code, module.__dict__)  # pylint: disable=exec-used
```
Comment (Corentin-pro) [inline ai/src/exec.py:19]: Using `module_spec.loader.exec_module(module)` seems better : this will not trigger unsafe exec warning from linter and won't run anything meant for manual run (with the `if __name__ == "__main__":` test).

File: ai/src/data.py [hunk 3]
```diff
@@ -238,3 +246,13 @@ def reward(
             raise RewardInvalidException(repr(ex)) from ex
 
         return loc["reward"]
+
+    def external_reward(self, action_name: str, prev_state_dict: dict, prev_state_interpretations,
```
Comment (Corentin-pro) [inline ai/src/data.py]: I would have get the method (`getattr`) directly in the initialization and put this function content (4 lines of code) directly in the `reward` function (avoiding passing all the parameters).

**Review by Corentin-pro — CHANGES_REQUESTED** [review]

File: ai/src/exec.py (hunk 2 above)
Comment (phillipleblanc) [reply ai/src/exec.py:19]: Hmm, I'm not specifying a loader here since there isn't a loader for loading from a string. I guess it would be possible to write `code` to a file and then use `spec_from_file_location`, but that seems non-optimal. Do you know of a way to get a loader that loads from a string?

Comment (Corentin-pro) [reply ai/src/exec.py:19]: After seeing `testutils-Snapshotter-SnapshotMultiJson-trader-external-reward.yaml.json` I see that you feed the code as a string in the manifest.

I was expecting the name of the file to load so people can code a python file directly, then we can use `importlib.util.spec_from_file_location` that will have a default loader.

Comment (phillipleblanc) [reply ai/src/exec.py:19]: Ideally, we'd have all of the filesystem interaction happen in the Go runtime. This allows us to keep the I/O interface for the Python to be only gRPC/HTTP APIs. The only filesystem interaction the Python project has right now is for saving and loading the keras models - which isn't ideal, but I wasn't able to find a way to keep the keras models loaded in memory and be able to inference on them in a multi-threaded way.

**Review by Corentin-pro — APPROVED** [review]
